File: tests/e2e/assert_firefox_container.py
#!/usr/bin/env python3
"""Assert one browser-produced Firefox container cookie through Python."""

from __future__ import annotations

import argparse
import sys

import rookie_cookies


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("database")
    parser.add_argument("user_context_id", type=int)
    args = parser.parse_args()
    snapshot = rookie_cookies.from_path(args.database)
    context = {
        "url": "https://container.rookie.test/",
        "top_level_site": "https://container.rookie.test/",
        "user_context_id": args.user_context_id,
    }
    if snapshot.header(context) != "rookie_container=container-1":
        raise RuntimeError("Python did not select the exact Firefox container cookie")
    if snapshot.header({**context, "user_context_id": args.user_context_id + 1}) != "":
        raise RuntimeError("Python merged a different Firefox container")
    try:
        snapshot.header(
            {
                "url": context["url"],
                "top_level_site": context["top_level_site"],
            }
        )
    except Exception as error:
        if getattr(error, "code", None) != "incomplete_send_context" and (
            "incomplete_send_context" not in str(error)
        ):
            raise
    else:
        raise RuntimeError("Python accepted a missing Firefox container selector")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"Firefox container Python assertion failed: {error}", file=sys.stderr)
        raise

File: tests/e2e/assert_firefox_container.mjs
// Assert one browser-produced Firefox container cookie through Node.

import process from "node:process";

import * as rookieCookies from "../../bindings/node/index.js";

const [database, idArg] = process.argv.slice(2);
const userContextId = Number(idArg);
if (!database || !Number.isSafeInteger(userContextId) || userContextId <= 0) {
  console.error(
    "usage: node assert_firefox_container.mjs <database> <user-context-id>",
  );
  process.exit(2);
}

const snapshot = await rookieCookies.fromPath({ path: database });
const context = {
  url: "https://container.rookie.test/",
  topLevelSite: "https://container.rookie.test/",
  userContextId,
};
if (snapshot.header(context) !== "rookie_container=container-1") {
  throw new Error("Node did not select the exact Firefox container cookie");
}
if (snapshot.header({ ...context, userContextId: userContextId + 1 }) !== "") {
  throw new Error("Node merged a different Firefox container");
}
try {
  snapshot.header({
    url: context.url,
    topLevelSite: context.topLevelSite,
  });
  throw new Error("Node accepted a missing Firefox container selector");
} catch (error) {
  if (
    error.rookieCode !== "incomplete_send_context" &&
    !String(error).includes("incomplete_send_context")
  ) {
    throw error;
  }
}
